built_value/lib/standard_json_plugin.dart

```diff
@@ -26,58 +33,170 @@ class StandardJsonPlugin implements SerializerPlugin {
+    var type = list[0];
+
+    // Length is at least two because we have one entry for type and one for
+    // the value.
+    if (list.length == 2) {
+      // Just a type and a primitive value. Encode the value in the map.
+      return <Object, Object>{discriminator: type, valueKey: list[1]};
+    }
+
+    if (type == 'list') {
+      // Embed the list in the map.
+      return <Object, Object>{discriminator: type, valueKey: list.sublist(1)};
+    }
+
+    // If a map has non-String keys then they need encoding to strings before
+    // it can be converted to JSON. Because we don't know the type, we also
+    // won't know the type on deserialization, and signal this by changing the
+    // type name on the wire to `encoded_map`.
+    var needToEncodeKeys = false;
+    if (type == 'map') {
+      for (int i = 0; i != (list.length - 1) ~/ 2; ++i) {
+        if (list[i * 2 + 1] is! String) {
+          needToEncodeKeys = true;
+          type = 'encoded_map';
+          break;
+        }
+      }
+    }
+
+    final result = <Object, Object>{discriminator: type};
+    for (int i = 0; i != (list.length - 1) ~/ 2; ++i) {
+      final key =
+          needToEncodeKeys ? _encodeKey(list[i * 2 + 1]) : list[i * 2 + 1];
+      final value = list[i * 2 + 2];
+      result[key] = value;
+    }
+    return result;
+  }
+
+  /// JSON-encodes an `Object` key so it can be stored as a `String`. Needed
+  /// because JSON maps are only allowed strings as keys.
+  String _encodeKey(Object key) {
     return JSON.encode(key);
   }
 
-  List _toList(Map map, bool alreadyStringKeys) {
+  /// Converts [StandardJsonPlugin] serialization output, a `Map`, to a `List`,
+  /// when the serialized type is known statically.
+  List _toList(Map map, bool hasEncodedKeys) {
     final result = new List(map.length * 2);
     var i = 0;
     map.forEach((key, value) {
       // Drop null values, they are represented by missing keys.
       if (value == null) return;
 
-      result[i] = alreadyStringKeys ? key : _fromStringKey(key as String);
+      result[i] = hasEncodedKeys ? _decodeKey(key as String) : key;
+      result[i + 1] = value;
+      i += 2;
+    });
+    return result;
+  }
+
+  /// Converts [StandardJsonPlugin] serialization output, a `Map`, to a `List`,
+  /// when the serialized type is not known statically. The type is retrieved
+  /// from the [discriminator] field.
+  List _toListUsingDiscriminator(Map map) {
+    var type = map[discriminator];
+
+    if (type == null) {
+      throw new ArgumentError('Unknown type on deserialization. '
+          'Need either specifiedType or discriminator field.');
+    }
+
+    if (type == 'list') {
+      return [type]..addAll(map[valueKey] as Iterable);
+    }
+
+    if (map.containsKey(valueKey)) {
+      // Just a type and a primitive value. Retrieve the value in the map.
+      final result = new List(2);
+      result[0] = type;
+      result[1] = map[valueKey];
+      return result;
```

Comment:
return [type, map[valueKey] ] ; ?

Reply:
I believe that would use more memory as it would return a growable list; as I have it, we get a fixed list of size two.